src/loaders/ncbi_downloader/gtdb.py

```diff
@@ -0,0 +1,311 @@
+
+    file_dir_url = (
+        GTDB_DOMAIN + f"release{release_ver}/" + f"{release_ver}.0/"
+    )  # TODO: add dot support, relase version 86 has '86.1' and '86.2'
+
+    file_urls = list()
+
+    # parse taxonomy file URL for genome ids
+    response = requests.get(file_dir_url)
+    soup = BeautifulSoup(response.text, "html.parser")
+
+    for link in soup.find_all("a"):
+        href = link.get("href")
+
+        if href.endswith("tsv") and "_taxonomy_" in href:
+            file_urls.append(file_dir_url + href)
+
+    if not file_urls or len(file_urls) >= 3:
+        raise ValueError(
+            f"Parsed unexpected taxonomy files {file_urls}. "
+            f"Please check parsing logic in _form_gtdb_taxonomy_file_url."
+        )
+
+    return file_urls
+
+
+def _fetch_gtdb_genome_ids(release_ver: str, work_dir: str) -> Tuple[list[str], list[str]]:
+    # download GTDB taxonomy files and then parse genome_ids from the GTDB taxonomy files
+    genome_ids = list()
+
+    taxonomy_urls = _form_gtdb_taxonomy_file_url(release_ver)
+
+    for taxonomy_url in taxonomy_urls:
+        # download GTDB taxonomy file to work_dir
+        url = urlparse(taxonomy_url)
+        taxonomy_file = os.path.join(work_dir, os.path.basename(url.path))
+        request.urlretrieve(taxonomy_url, taxonomy_file)
+
+        # parse genome id from first column of GTDB taxonomy tsv file
+        with open(taxonomy_file, "r") as f:
+            genome_ids.extend(
+                [parse_genome_id(line.strip().split("\t")[0]) for line in f]
+            )
+
+    return genome_ids, taxonomy_urls
+
+
+def _make_work_dir(root_dir: str, source_data_dir: str, source: str, env: str) -> str:
+    # make working directory for a specific collection under root directory
+    work_dir = os.path.join(root_dir, source_data_dir, source, env)
+    os.makedirs(work_dir, exist_ok=True)
+    return work_dir
+
+
+def _process_genome_ids(
+    work_dir: str,
+    genome_ids_unprocssed: list[str],
+    target_file_ext: list[str],
+    exclude_name_substring: list[str],
+    overwrite: bool = False,
+) -> list[str]:
+    """
+    Helper function that processes genome ids to avoid redownloading files.
+    """
+    if overwrite:
+        return genome_ids_unprocssed
+
+    genome_ids = list()
+    target_ext_count = len(target_file_ext)
+    for genome_id in genome_ids_unprocssed:
+        data_dir = os.path.join(work_dir, genome_id)
+        if not os.path.exists(data_dir):
+            genome_ids.append(genome_id)
+            continue
+
+        ext_count = 0
+        for file_name in os.listdir(data_dir):
+            if any([file_name.endswith(ext) for ext in target_file_ext]) and all(
+                [substring not in file_name for substring in exclude_name_substring]
+            ):
+                ext_count += 1
+
+        if ext_count != target_ext_count:
+            genome_ids.append(genome_id)
+
+    return genome_ids
+
+
+def main():
```

Comment:
main is 137 lines which is way too long, and needs some splitting up. Arg parsing is one sub function that could be split out, for example

Reply:
Updated!

Reply:
Still kind of long but moving the threading code to the ncbi helper script will help so for now :+1:

Reply:
Updated!